sklearn/tree/_partitioner.pyx

```diff
@@ -0,0 +1,832 @@
+"""Partition samples in the construction of a tree.
+
+This module contains the algorithms for moving sample indices to
+the left and right child node given a split determined by the
+splitting algorithm in `_splitter.pyx`.
+
+Partitioning is done in a way that is efficient for both dense data,
+and sparse data stored in a Compressed Sparse Column (CSC) format.
+"""
+# Authors: The scikit-learn developers
+# SPDX-License-Identifier: BSD-3-Clause
+
+from cython cimport final
+from libc.math cimport isnan, log
+from libc.stdlib cimport qsort
+from libc.string cimport memcpy
+from scipy.sparse import issparse
+
+import numpy as np
+
+
+# Constant to switch between algorithm non zero value extract algorithm
+# in SparsePartitioner
+cdef float32_t EXTRACT_NNZ_SWITCH = 0.1
+
+# Allow for 32 bit float comparisons
+cdef float32_t INFINITY_32t = np.inf
+
+
+cdef class Partitioner:
```

Comment:
According to your opening comment, is this suppose to be `BasePartitioner`?

Also, is defining this base class just for defining the interface? I didn't define the interface so we can be sure there is no v-table look up. I do not think Cython would generate one, but there could be a bug in Cython that will generate it in the future, and those are hard to detect.

If we want to be safe, we can comment out the interface and let it be a comment.

Reply:
> According to your opening comment, is this suppose to be `BasePartitioner`?

Yes, I should change it back.

> Also, is defining this base class just for defining the interface? I didn't define the interface so we can be sure there is no v-table look up. I do not think Cython would generate one, but there could be a bug in Cython that will generate it in the future, and those are hard to detect.

Yes it's just to define the interface, but I noticed no runtime regression. Just to confirm, your suggestion is to remove the abstract interface definition, and just leave one there as a multi-in-line comment?

Reply:
> Just to confirm, your suggestion is to remove the abstract interface definition, and just leave one there as a multi-in-line comment?

Yea, that was my suggestion.

Reply:
Kay I have done that in [e24cda9](https://github.com/scikit-learn/scikit-learn/pull/29458/commits/e24cda96ae85f5200e6cd85e546e23ee98a9427a)